### PR TITLE
multi address for rpc

### PR DESCRIPTION
**Pull request by xrdavies**
* add method xdag_new_address for json-rpc
* add feature multi address for wallet, need pool support

### CODE REVIEW

**Comment by xrdavies** [comment]
only available with rpc enabled, and use command -multi to enable multi address for wallet and pool.

File: client/json-rpc/rpc_procedures.c [hunk 1]
```diff
@@ -496,6 +500,60 @@ cJSON * method_xdag_get_transactions(struct xdag_rpc_context * ctx, cJSON * para
 	return result;
 }
 
+/* create new addresss */
+/*
+ request:
+ "method":"xdag_new_address", "params":[{"number":1}], "id":1
```
Comment (jonano614) [inline client/json-rpc/rpc_procedures.c]: Could you please clarify params and result.

File: client/miner.c [hunk 2]
```diff
@@ -139,6 +159,89 @@ static int send_to_pool(struct xdag_field *fld, int nfld)
 	return 0;
 }
 
+static int miner_fresh_callback(void *data, xdag_hash_t hash, xdag_amount_t amount, xdag_time_t time, int n_our_key)
```
Comment (jonano614) [inline client/miner.c]: data, amount, time and n_our_key  are unused

File: client/json-rpc/rpc_procedures.c (hunk 1 above)
Comment (xrdavies) [reply client/json-rpc/rpc_procedures.c]: Number is the number of address to be generated.

Result is the array of generated addresses.

File: client/miner.c (hunk 2 above)
Comment (xrdavies) [reply client/miner.c]: Yes, they are required by xdag_traverse_our_blocks, but not used directly.

File: client/json-rpc/rpc_procedures.c (hunk 1 above)
Comment (jonano614) [reply client/json-rpc/rpc_procedures.c]: Do you mean count of addresses?

File: client/miner.c (hunk 2 above)
Comment (jonano614) [reply client/miner.c]: So, maybe it makes sense to remove them?

File: client/miner.c [hunk 3]
```diff
@@ -139,6 +159,89 @@ static int send_to_pool(struct xdag_field *fld, int nfld)
...
+				xdag_err("socket is closed");
+				pthread_mutex_unlock(&g_miner_mutex);
+				break;
+			}
+			pthread_mutex_unlock(&g_miner_mutex);
+
+			sleep(0.5); // sleep half second
+		}
+
+		sleep(120); // 5 mintues
```
Comment (jonano614) [inline client/miner.c]: minutes

File: client/miner.c [hunk 4]
```diff
@@ -197,6 +300,22 @@ void *miner_net_thread(void *arg)
 		pthread_mutex_unlock(&g_miner_mutex);
 		goto err;
 	}
+
+	if(g_multi_address){
```
Comment (jonano614) [inline client/miner.c]: I think it is not a good idea to start the thread here. I suggest you to do it in xdag_initialize_miner function

File: client/miner.c [hunk 5]
```diff
@@ -278,7 +397,33 @@ void *miner_net_thread(void *arg)
...
-					maxndata = 2 * sizeof(struct xdag_field);
+
+					int has = 0;
+
+					if(g_multi_address) {
+						pthread_mutex_lock(&g_mutex_our_block_list);
+						struct block_list_element * elem = NULL;
+						LL_FOREACH(g_our_blocks_list, elem)
+						{
+							if(memcmp(last->data, elem->hash, sizeof(xdag_hashlow_t)) == 0) {
```
Comment (jonano614) [inline client/miner.c]: That part of code is duplicated. I think we can move it to separate function that returns 1 if memcmp == 0 and 0 in other case.

File: client/miner.h [hunk 6]
```diff
@@ -19,6 +19,11 @@ extern int xdag_mining_start(int n_mining_threads);
 /* initialization of connection the miner to pool */
 extern int xdag_initialize_miner(const char *pool_address);
 
+/* create new wallet address */
+extern int xdag_miner_new_address(xdag_hash_t hash);
+
+extern void *miner_refresh_thread(void *arg);
```
Comment (jonano614) [inline client/miner.h]: This declaration is not necessary in the header. Function is used only in miner.c file.

File: client/pool.c [hunk 7]
```diff
@@ -89,6 +94,7 @@ struct connection_pool_data {
 	uint64_t nfield_out;
 	uint64_t task_index;
 	struct xdag_block *block;
+	struct wallet_list_element * wallet_address_list;
```
Comment (jonano614) [inline client/pool.c]: I can hardly say I like that change... But ok, we can temporary do it. Until RPC for wallets is ready.

File: client/pool.c [hunk 8]
```diff
@@ -772,6 +788,30 @@ static int is_block_data_received(connection_list_element *connection)
 				}
 
 				pthread_mutex_unlock(&g_pool_mutex);
+
+				if(g_multi_address) {
```
Comment (jonano614) [inline client/pool.c]: What will happen if wallet is multi addresses connect to pool without activated multi addresses? And vice versa.

Comment (jonano614) [reply client/pool.c]: What will happen if transaction will be sent?

File: client/pool.c [hunk 9]
```diff
@@ -938,6 +978,35 @@ static int send_data_to_connection(connection_list_element *connection, int *pro
...
+				memcpy(balance.data, elem->hash, sizeof(xdag_hash_t));
+				balance.amount = xdag_get_balance(balance.data);
+
+				char buf[33];
+				xdag_hash2address(balance.data, buf);
+				xdag_mess("send balance for %s", buf);
+
+				LL_DELETE(conn_data->wallet_address_list, elem);
+
+				dfslib_encrypt_array(g_crypt, (uint32_t*)&balance, DATA_SIZE, conn_data->nfield_out++);
```
Comment (jonano614) [inline client/pool.c]: This part of logic is duplicated and can be moved to separa fuction.

File: client/pool.c (hunk 7 above)
Comment (xrdavies) [reply client/pool.c]: Would say until full node is ready. 😃

File: client/pool.c (hunk 8 above)
Comment (xrdavies) [reply client/pool.c]: If pool does not enable multi addresses, this package will be no processed.

File: client/json-rpc/rpc_procedures.c (hunk 1 above)
Comment (xrdavies) [reply client/json-rpc/rpc_procedures.c]: Good idea

File: client/miner.c (hunk 3 above)
Comment (xrdavies) [reply client/miner.c]: Ah, typo

File: client/miner.c (hunk 4 above)
Comment (xrdavies) [reply client/miner.c]: Good suggestion, I will move to xdag_initialize_miner

**Comment by xrdavies** [comment]
@jonano614 @sgaragagghu My final decision is to close this PR due to no solution to solve the security issue of mutli-address. Prefer to use other options to satisfy the requirements from exchanges.

**Pull request closed without merging**